BLD: Add support for building iOS wheels

freakboy3742 opened the pull request:
Adds iOS as a test and release target for NumPy.

**UPDATE: 17 July 2026** This PR is now ready for review. PR body reflects current status. 

In an environment where:
* Xcode and the iOS SDK is installed, with all licences accepted.
* cibuildwheel 4.1.0 is installed
* ninja is available on the path

This branch will build and test accelerated iOS simulator wheels:
```
CIBW_PLATFORM=ios CIBW_ENVIRONMENT="IPHONEOS_DEPLOYMENT_TARGET=17.0" cibuildwheel --arch arm64_iphonesimulator
```
You can also specify `arm64_iphoneos` for device wheels, or `x86_64_iphonesimulator` for simulator wheels on older hardware.

The build will also work without the `IPHONEOS_DEPLOYMENT_TARGET` definition, but with the older LP64 version of accelerate (this is the analog of the `MACOSX_DEPLOYMENT_TARGET=14.0` configuration, except that there isn't a pre-compiled openBLAS for iOS to fall back on). To build with a lower minimum iOS version, run:
```
CIBW_PLATFORM=ios CIBW_BUILD="cp3*" CIBW_CONFIG_SETTINGS="setup-args=-Duse-ilp64=false setup-args=-Dallow-noblas=true build-dir=build" cibuildwheel --archs arm64_iphonesimulator
```

freakboy3742 committed:
Add support for building iOS wheels.

freakboy3742 committed:
Encode the iOS longdouble_format.

freakboy3742 commented:
And update; after discussions on the mesonbuild/meson-python#731, some additional changes to meson (mesonbuild/meson#14541) and CPython (python/cpython#133184) were identified. I'm working on getting those changes upstreamed.

henryiii commented:
cibuildwheel 3.0 out. :)

freakboy3742 commented:
> cibuildwheel 3.0 out. :)

Awesome! A new meson-python release is only remaining blocker.

freakboy3742 committed:
Drop 3.14 support from iOS CI until cibuildwheel is available.

freakboy3742 committed:
Use main branch of meson-python

freakboy3742 committed:
Use a better check for availability of subprocesses.

freakboy3742 committed:
Modify test suite to support iOS.

freakboy3742 commented:
Status update: With the current main branch of meson-python, numpy/meson#22 applied to the meson-python branch, and the current main branch of cibuildwheel, this passes the full test suite on iOS.

The changes to the test suite changes involve:
1. making test skips that were WASM-specific into a more generic test skip when subprocess isn't available, or 
2. making known macOS-specific test skips generic to *all* Apple platforms.

On top of those changes, there's one specific change to ctypes handling to account for the odd way that iOS handles dynamic loading, and one workaround for a pytest issue (related to getuser() and the creation of temporary files).

freakboy3742 commented on numpy/testing/_private/utils.py:
```diff
@@ -99,6 +103,13 @@ class KnownFailureException(Exception):
+# getuser() tries to use `pwd`, which doesn't exist on iOS. Fake a
+# username that will take priority.
+if IS_IOS:
+    os.environ["USER"] = "iOSUser"
```
This workaround isn't required if you use pytest 8.1.0 or higher, as the underlying problem was fixed by pytest-dev/pytest#11875

freakboy3742 replied:
I've removed this workaround by bumping pytest to 8.4.1.

freakboy3742 committed:
Add comment about the getuser() workaround.

freakboy3742 committed:
Bump vendored meson version.

freakboy3742 committed:
Bump pytest version to remove need for 3.13 username workaround.

freakboy3742 committed:
Correct linting issues.

freakboy3742 committed:
Minor cleanup of wheel building matrix.

freakboy3742 committed:
Test fixes masked by iOS platform exclusions.

freakboy3742 committed:
Add iOS test workflow, and bump cibuildwheel to 3.1.2.

freakboy3742 committed:
Correct variable name in iOS workflow.

freakboy3742 committed:
Remove use of macos-13 from iOS workflows.

freakboy3742 commented:
I've added CI workflows, and updated cibuildwheel and vendored-meson versions; the meson-python release is the only outstanding TODO.

I've updated the description of the PR to reflect current status, and a couple of notable design decisions.

freakboy3742 committed:
Merge branch 'main' into ios-support

freakboy3742 committed:
Add typing info for some new constants.

freakboy3742 committed:
Update CI configurations.

freakboy3742 committed:
Correct quoting on matrix shell check.

freakboy3742 committed:
Break out iOS workflows to one arch per build.

mhsmith commented on .github/workflows/wheels.yml:
```diff
@@ -93,12 +96,24 @@ jobs:
+        run: |
+          if [[ "${{ matrix.buildplat[3] }}" == 'accelerate' ]]; then
+            # Accelerate with ILP64 requires iOS 16.4 or higher
+            CIBW="IPHONEOS_DEPLOYMENT_TARGET=17.0 INSTALL_OPENBLAS=false RUNNER_OS=iOS"
```
iOS 17 is only 2 years old, so this means a lot of apps, including the default Briefcase configuration, will end up using the unaccelerated build, which may be ["100x slower"](https://github.com/numpy/numpy/blob/v2.3.5/numpy/meson.build#L185-L187). 

Could we avoid this by building in non-ILP64 mode? As I understand it, the only advantage of ILP64 is that it allows arrays with more than 2 billion elements, which, on a phone, is far less likely to be a problem than a 100x slowdown. In that case, there probably isn't even any need for a separate iOS 17.0 build, and the workflows can all be simplified.

freakboy3742 replied:
It may be "only" 2 years old, but a feature of the iOS ecosystem is by mass and near-immediate adoption of OS updates. Based on [this chart](https://telemetrydeck.com/survey/apple/iOS/majorSystemVersions/), as of today, 97.5% of devices are on iOS 17 or newer. I included the "non accelerate" wheels mostly as a convenience so that a a default iOS install will have access to *something*.

I can take a look to see if it's possible to pick up the older accelerate implementation on iOS13+.

freakboy3742 commented on numpy/_core/meson.build:
```diff
@@ -439,7 +439,11 @@ endforeach
-  longdouble_format = meson.get_compiler('c').run(
+  # We can't programmatically check iOS; but we do know what format it uses.
+  if host_machine.system() == 'ios'
+    longdouble_format = 'IEEE_DOUBLE_LE'
```
Agreed that a link to docs is worthwhile here.

As for the 80-bit float... I guess we could add a special case for x86-64 here, but my immediate thought is whether it's preferable to retain size compatibility with ARM64 *hardware*, on the basis that you really want to retain the performance characteristics that you're going to see on physical devices, rather than retain the full capabilities of the simulator platform. 

(Plus... x86-64 simulator support has less than 2 years of life left, so maintain extensive compatibilities considerations seems overkill)

mhsmith replied:
Agreed we don't need to go out of our way to worry about x86_64 performance, but this looks like it might be the kind of setting where you must match the format expected by the compiler and the BLAS library, or else risk a low-level crash or completely incorrect calculations. 

On the other hand, if it's passing the full NumPy test suite on both architectures, then I guess that proves it's OK.

freakboy3742 committed:
Remove misleading reference to OpenBLAS.

freakboy3742 committed:
Add clarifying note about iOS longdouble formats.

freakboy3742 committed:
Merge branch 'main' into ios-support

rgommers commented:
> A new meson-python release is only remaining blocker.

Sorry that that took so long 🙄. `meson-python` 0.19.0 is up on PyPI now, conda-forge will follow shortly.

freakboy3742 committed:
Merge branch 'main' into ios-support

freakboy3742 committed:
Bump meson-python version.

freakboy3742 committed:
Revert ruff formatting.

freakboy3742 committed:
Add extra iOS test skip.

freakboy3742 committed:
Correct linting error.

freakboy3742 committed:
Revert *all* the ruff changes.

mhsmith commented:
@freakboy3742: What's the status of this PR?

freakboy3742 committed:
Merge branch 'main' into ios-support

freakboy3742 committed:
Update tests and CI config.

freakboy3742 committed:
Drop x86-64 wheels, and bump test dependency versions.

freakboy3742 committed:
Update build instructions to generate Accelerate builds for older iOS versions.

freakboy3742 commented:
My apologies - this fell off my radar for a bit; I was waiting for a cibuildwheel release for some patches. Those changes have now landed, and I've merged this PR with the current main.

The PR could be reviewed (and potentially merged) as is as of 11c74b72 (the most recent-but-one commit).

That build removes x86-64 simulator wheels from the build. In the [most recent CI passes](https://github.com/numpy/numpy/actions/runs/24427249397/job/71363874546), the wheels build and pass *almost* all the tests, but here were failing some tests verifying the print format of floating point and complex types. I'm guessing they're related to the [odd float size](https://github.com/numpy/numpy/pull/28759#discussion_r2605602156)I couldn't find an immediately obvious cause for these problems (they're likely related to the odd float size, so I came to the conclusion that it wasn't worth the effort. 

The x86-64 simulator platform is on borrowed time at this point. Apple no longer produced x86-64 hardware; Tahoe (macOS 26, released last year) is the last macOS that will include x86-64 support; Circle CI dropped x86-64 simulators 2 years ago, and Github has indicated that they're dropping support in Fall of next year. So - I'd need to do a bunch of debugging for a hardware platform that we'd need to deprecate in just over 12 months. 

The most recent commit is failing CI on older iOS versions. There was a [discussion above](https://github.com/numpy/numpy/pull/28759#discussion_r2605678212) about whether we should be using "non-accelerated" iOS wheels, or using the older Accelerate interface; I've done some tinkering, and with the inclusion of numpy/meson#25 in the vendored meson, I can generate "pre-NEW_LAPACK" wheels for iOS 13.0.

rgommers commented:
> so I came to the conclusion that it wasn't worth the effort.

That sounds right to me.

> with the inclusion of [numpy/meson#25](https://github.com/numpy/meson/pull/25) in the vendored meson, I can generate "pre-NEW_LAPACK" wheels for iOS 13.0.

That looks pretty clean, I'd be fine with merging that if it's needed. I'll have a look at that tomorrow.

freakboy3742 committed:
Merge branch 'main' into ios-support

freakboy3742 committed:
Updates for vendored meson changes.

rgommers commented:
The iOS Accelerate support is in now (xref gh-32020), apologies again for the delay. 

This overlaps with gh-30412, so let's get that merged before fixing the merge conflicts here I'd say.

rgommers commented:
gh-30412 is now merged, so this PR should be unblocked.

freakboy3742 committed:
Merge branch 'main' into ios-support

freakboy3742 committed:
Simplify actions workflows.

freakboy3742 committed:
Clean up linting issues from merge.

freakboy3742 committed:
Refresh iOS workflow version pins.

freakboy3742 committed:
Ignore local .envrc files.

freakboy3742 committed:
Correct an iOS config setting.

freakboy3742 committed:
Add an xfail for potentially thread unsafe operation on Android.

freakboy3742 committed:
Try x86-64 wheels again.

freakboy3742 commented on numpy/_core/tests/test_multithreading.py:
```diff
@@ -89,8 +89,22 @@ def test_eigvalsh_thread_safety():
+                "version with a fix (0.3.33.112) or BLAS metadata is not "
+                "sufficient to identify the BLAS implementation."
+            )
+        raise
```
This modification isn't iOS specific - I've added it because of [this build failure on Android](https://github.com/numpy/numpy/actions/runs/29545977646/job/87778353719).

(Tagging @mhsmith for visibility)

freakboy3742 committed:
Correct longdouble format for x86-64 simulators.

freakboy3742 committed:
Disable x86-64 in CI - takes too long to run.

rgommers commented on .github/workflows/ios.yml:
```diff
@@ -0,0 +1,72 @@
+            accelerated: true
+            arch: arm64_iphonesimulator
+
+          # Non-accelerated
```
This is no longer true, right? It will call Accelerate with non-suffixed symbols now.

freakboy3742 replied:
True - I'll clarify the comment.

rgommers commented on .github/workflows/ios.yml:
```diff
@@ -0,0 +1,72 @@
+            echo "CIBW_ENVIRONMENT_IOS=$CIBW_ENV" >> "$GITHUB_ENV"
+          else
+            # On older iOS versions, disable ILP64.
+            CIBW_CONFIG="setup-args=-Duse-ilp64=false setup-args=-Dallow-noblas=true build-dir=build"
```
Should be `Dallow-noblas=false` now, right?

freakboy3742 replied:
Correct. Update incoming

rgommers commented on numpy/_core/tests/test_multithreading.py:
```diff
@@ -89,8 +89,22 @@ def test_eigvalsh_thread_safety():
+        run_threaded(lambda i: np.linalg.eigvalsh(matrices[i]), 2,
+                     pass_count=True)
+    except np.linalg.LinAlgError:
+        # A non-thread-safe LAPACK can corrupt the concurrent eigvalsh calls,
```
This looks like a change that's unrelated to iOS introduction, perhaps picked up in a merge commit?

freakboy3742 replied:
It is; as mentioned in the inline comment, I saw a test failure in Android related to this. However, I can pull this into a separate PR.

rgommers replied:
Ah sorry, I missed that comment, resolving. Might be better as a separate PR indeed, for visibility - the comment on this PR is likely to get lost and someone might go digging for where this actually happened.

freakboy3742 replied:
I've opened #32028 to track the issue. Looking closer, I'm not *completely* convinced my original fix was correct - it certainly would have silenced the problem, but possibly for the wrong reason. The problem hasn't occurred again in subsequent CI runs.

I've referenced the skip option on the ticket; I'll leave it up to someone more familiar with that corner of the code to make a call on whether a skip is called for.

rgommers replied:
Sounds good, I'll resolve this comment.

rgommers commented on pyproject.toml:
```diff
@@ -208,6 +208,14 @@ build-dir = "build"
 
+[tool.cibuildwheel.ios]
+# The ILP64 version of Accelerate is only available on iOS 16.4 or above.
+config-settings = "setup-args=-Duse-ilp64=true setup-args=-Dallow-noblas=true build-dir=build"
```
Same here, this now picks up Accelerate on older iOS versions, so I think you'd want `-Dallow-noblas=false` to not have it potentially go missing silently.

rgommers commented on pyproject.toml:
```diff
@@ -247,7 +255,7 @@ test-requires = [
 # run Python directly.
 test-command = """
-    python -m pytest --pyargs numpy \
+    python -m pytest --pyargs numpy -vv \
```
This needs to be reverted before merge I think? The CI logs are very verbose and are getting truncated.

freakboy3742 replied:
I prefer using -vv in CI config; with the added benefit that on iOS, stdout.flush() is interpreted as a newline, so you get lots of `.` on a single line in the logs. However, I'm happy to go with project preference here.

rgommers replied:
Thanks. Yes, we prefer without - the logs opened slowly for me already, and were truncated.

freakboy3742 replied:
Done.

rgommers commented on pyproject.toml:
```diff
@@ -208,6 +208,14 @@ build-dir = "build"
+# The ILP64 version of Accelerate is only available on iOS 16.4 or above.
+config-settings = "setup-args=-Duse-ilp64=true setup-args=-Dallow-noblas=true build-dir=build"
+# On older iOS versions, use the following configuration
+# config-settings = "setup-args=setup-args=-Duse-ilp64=false setup-args=-Dallow-noblas=true build-dir=build"
```
`-Duse-ilp64=false` is the default, so can be left off. I think you want simply:
```
config-settings = "setup-args=-Dallow-noblas=false build-dir=build"
```

rgommers commented on .github/workflows/wheels.yml:
```diff
@@ -47,6 +47,8 @@ jobs:
           # runner. So we build for ARM on a non-ARM runner, which will skip the tests.
           - [ubuntu-24.04, android_arm64_v8a]
           - [ubuntu-24.04, android_x86_64]
+          - [macos-26, ios_arm64_iphoneos, ""]
```
This is too much, let's remove this one.

rgommers reviewed (commented):
This looks very close, only a few minor things in the diff. The main remaining issue is the CI config: you're adding 6 jobs here, which is fine for testing on this PR, but it's way too much to merge - I already see the CI queue backing up because of the limited macOS runner concurrency. Two points
1. There's no need to duplicate testing of the same platform/blas config between `ios.yml` and `wheels.yml`. The former even invokes `cibuildwheel` anyway, so it looks like a 100% duplication.
2. I wouldn't test the path with no Accelerate linkage in CI, that's of very limited real-world value and pretty unlikely to regress anyway.

It looks to me like two CI jobs is the way to go here: one for `arm64_iphoneos` and one for `arm64_iphonesimulator`. Two `wheels.yml` entries would cover that, meaning the `ios.yml` can be removed completely, unless I am missing something here.

freakboy3742 commented:
I've merged with main as of today - and this PR is finally ready for review 🎉 

There's one fix in this PR that isn't related to iOS - I got a test failure on Android that appears to be caused by a multithreading issue with OpenBlas - see the [comment inline](https://github.com/numpy/numpy/pull/28759#discussion_r3599999242).

I've also been able to fix the issue with x86-64 simulator support. The downside: the x86-64 simulator CI pass is *slow* - it takes 40+ minutes to run the test suite. This is a combination of a [known issue with GitHub Actions](https://github.com/actions/runner-images/issues/12545), and the fact that x86-64 macOS hardware *is* older and slower. As a result, a CI pass running python 3.13, 3.14 and 3.15 takes over 2 hours.

That's obviously a significant choke point on limited macOS runners. Given that x86-64 support on macOS (and thus iOS) is deprecated and near EOL anyway, I've opted to keep the *code* to support x86-64 iOS simulators in place (as it was essentially 2 lines of Meson config), but removed the CI configuration.

rgommers commented:
Our comments just crossed:)

> I've opted to keep the _code_ to support x86-64 iOS simulators in place (as it was essentially 2 lines of Meson config), but removed the CI configuration.

This sounds like the right call. We have support for lots of legacy and niche platforms, and we'll continue to accept patches as needed, but we indeed can't cover everything in CI.

freakboy3742 committed:
Remove a CI target.

freakboy3742 committed:
Clarify allow-noblas usage.

freakboy3742 committed:
Clarify status of accelerate.

freakboy3742 committed:
Revert the Android test change.

mattip commented:
> I got a test failure on Android that appears to be caused by a multithreading issue with OpenBlas - see the https://github.com/numpy/numpy/pull/28759#discussion_r3599999242.

We are at 0.3.33.201, and OpenBLAS [just released 0.3.24](https://github.com/OpenMathLib/OpenBLAS/releases), I will be updating our build soon to 0.3.24. What do I need to do to get [scipy-openblas](https://github.com/MacPython/openblas-libs) to build the needed wheels?

freakboy3742 committed:
Correct envvar naming.

freakboy3742 commented:
> > I got a test failure on Android that appears to be caused by a multithreading issue with OpenBlas - see the [#28759 (comment)](https://github.com/numpy/numpy/pull/28759#discussion_r3599999242).
> 
> We are at 0.3.33.201, and OpenBLAS [just released 0.3.24](https://github.com/OpenMathLib/OpenBLAS/releases), I will be updating our build soon to 0.3.24. What do I need to do to get [scipy-openblas](https://github.com/MacPython/openblas-libs) to build the needed wheels?

I'll have to defer to @mhsmith on that one - the OpenBLAS package is [currently being installed from a ChaquoPy wheel store](https://github.com/numpy/numpy/blob/f67f65a1ab0d0998896703d6f205912d6975f966/tools/wheels/cibw_before_build.sh#L38-L42), providing 0.3.33. I'm not familiar with the process for publishing to that store.

freakboy3742 commented:
... and we have a green board! 

@rgommers I think I've addressed all your comments; let me know if you need anything else to land this!

rgommers commented:
Thanks @freakboy3742. The last commits all look good. The main remaining thing is the one in [my review comment above](https://github.com/numpy/numpy/pull/28759#pullrequestreview-4719521684). You still have 5 CI jobs here, including 2 quite slow ones. I think a sparse test matrix of 2 jobs is appropriate at this point.

freakboy3742 committed:
More simplification of iOS test matrix.

freakboy3742 commented:
> Thanks @freakboy3742. The last commits all look good. The main remaining thing is the one in [my review comment above](https://github.com/numpy/numpy/pull/28759#pullrequestreview-4719521684). You still have 5 CI jobs here, including 2 quite slow ones. I think a sparse test matrix of 2 jobs is appropriate at this point.

My apologies - between the overlapped responses I misunderstood the request on that one. 

I've now removed the `ios` workflow, and modified the `wheels` workflow to run 2 jobs - one sim, one device, both arm64 and ILP64 ACCELERATE_NEW_LAPACK enabled.

rgommers reviewed (approved):
Awesome, in it goes Thanks again @freakboy3742!
